client/my-sites/customer-home/locations/primary/index.jsx

```diff
@@ -46,7 +47,7 @@ const cardComponents = {
 
 const Primary = ( { cards } ) => {
 	if ( ! cards || ! cards.length ) {
-		return null;
+		return <PerformanceTrackerStop id={ 'home' } />;
```

Comment:
To prevent duplication and avoid bugs (what if another early `return` is added to the component?), component composition can be helpful:
```jsx
const PrimaryContent = ( { cards } ) => {
  if ( ! cards ) {
    return null;
  }
  ...
};

const Primary = ( ...props ) => {
  return (
    <>
      <PrimaryContent { ...props } />
      <PerformanceTrackerStop id="home" />
    </>
  );
};
```

This wrapping can be made more generic by turning it into a HoC:
```jsx
withPerformanceTrackerStop( 'home' )( Primary );
```
or it can be a hook:
```js
const primary = ( { cards } ) => {
  usePerformanceTrackerStop( 'home' );
  ...
}
```
It doesn't need to be a component: it doesn't render anything and hooks are sufficient for declaring effects. The only limitation is that the hook can't be used conditionally.

Reply:
Good idea!

Reply:
I'm not sure about the HoC approach. Having the instrumentation "far away" from the content rendered can result in measuring components that are still loading. For example, if we have a component like:

```js
const Primary = () => {
	return <PrimaryContent />;
};

// many lines later
withPerformanceTrackerStop( 'home' )( Primary );
```

it may happen that someone adds a loading state and forgets to "exclude" it from the tracking. For example, someone could apply this change:

```diff
-const Primary = ( ) => {
+const Primary = ( { isLoading } ) => {
+	if ( isLoading ) return <PrimaryLoading />;
	return <PrimaryContent />;
};
```

and not realize that the performance tracker doesn't include the content anymore.

Of course this is a potential problem in any case (someone my forget to change `<PerformanceTrackerStop id="home" />`), but I guess it will be a bit harder to miss it bug if the component is explicitly used inside the rendering function and not from a composed component.

Do you think that could happen? How can we mitigate it?

Reply:
> it may happen that someone adds a loading state and forgets to "exclude" it from the tracking.

It depends on what exactly you want to measure. Your original code stopped the tracker even for the case where the cards are not loaded or loading (`( ! cards || ! cards.length )`). Here I'm not sure whether the condition means "cards are loading" or "there are zero cards". Need to be careful about this ambiguity.

If you want to stop the tracker only when the content is truly loaded (from REST), then the component should be rendered only in the "render loaded content" branch, of course.

Also, the original code would lead to the `stop` function being called twice: once from the `PerformanceTrackerStop` instance in the `( ! cards )` branch, and then, once loaded, from the "render content" branch. The `stop` function is probably resilient against that, but good to be aware of it.